Desktop_Interface/functiongencontrol.cpp

```diff
@@ -86,65 +83,103 @@ void functionGenControl::waveformName(QString newName, ChannelData& channel, int
-        channel.samples[i] = (unsigned char) dummy;
+        channel.samples[i] = uint8_t(dummy);
     }
 
 #else
-    QString directory = QCoreApplication::applicationDirPath();
 
-    directory.append("/waveforms/");
-    directory.append(newName);
-    QByteArray temp = encode(directory);
-    char *fileName = temp.data();
+    QByteArray filePath = QCoreApplication::applicationDirPath()
+		.append("/waveforms/").append(newName).toUtf8();
 
-    qDebug() << "opening" << fileName;
-    FILE *fptr = fopen(fileName, "r");
-        if (fptr==NULL){
-            qFatal("%s could not be opened!", fileName);
-        }
+    qDebug() << "opening" << filePath;
 
-    char lengthString[LengthStringSize]; // LENGTH 6
+    FILE *fptr = fopen(filePath.constData(), "r");
+    if (fptr == NULL)
+        qFatal("%s could not be opened!", filePath.constData());
+
+    char lengthString[16];
     fgets(lengthString, 5, fptr);
     sscanf(lengthString, "%d", &channel.length);
 
-    char divisibilityString[LengthStringSize]; // LENGTH 6
+    char divisibilityString[16];
     //Bit of bullshit to deal with CRLF line endings on Mac.
-    do fgets(divisibilityString, 5, fptr);
-    while ((divisibilityString[0] == '\r') || (divisibilityString[0] == '\n'));
+    do
+	{
+		fgets(divisibilityString, 5, fptr);
+	}
+	while ((divisibilityString[0] == '\r') || (divisibilityString[0] == '\n'));
+
     sscanf(divisibilityString, "%d", &channel.divisibility);
 
     qDebug() << "Length = " << channel.length;
     qDebug() << "Divisibility = " << channel.divisibility;
 
-    free(channel.samples);
-    channel.samples = (unsigned char *) malloc(channel.length);
+	channel.samples.resize(channel.length);
 
     char *dataString = (char *) malloc(channel.length*5+1);
     fgets(dataString, channel.length*5+1, fptr);
 
     int dummy;
     char *dataStringCurrent = dataString;
-    for (int i=0;i<channel.length;i++){
+    for (int i = 0; i < channel.length; i++)
+	{
         sscanf(dataStringCurrent, "%d", &dummy);
         dataStringCurrent += strcspn(dataStringCurrent, "\t") + 1;
-        channel.samples[i] = (unsigned char) dummy;
+        channel.samples[i] = uint8_t(dummy);
     }
 
     free(dataString);
     fclose(fptr);
 #endif
 
-    setMaxFreq_CH1(DAC_SPS/(channel.length>>(channel.divisibility-1)));
-    setMinFreq_CH1((double) CLOCK_FREQ/1024/65535/channel.length);
-    functionGenToUpdate(channelID, this); // 1ST ARG = 0
+	double newMaxFreq = DAC_SPS / (channel.length >> (channel.divisibility - 1));
+	double newMinFreq = double(CLOCK_FREQ) / 1024.0 / 65535.0 / channel.length;
+
+	// NOTE: Not very clean... Not sure what to do about it.
+	// I guess the "right thing" would be to have a Channel QObject class with its
+	// own signals and slots, or have a single setMaxFreq signal with channelID as
+	// an argument. Either solution would require changes in other places in the
+	// codebase so this will have to do for now.
```

Comment:
Haha, that honestly that isn't the worst thing in the world.  Yes, it's an unnecessary if-else block but there's so much worse on this Earth.